eos-boot-upgrade

```diff
@@ -0,0 +1,278 @@
+        if dir_cmp(sub_dcmp):
+            return True
+    return False
+
+
+def check_for_upgrade(esp_mount, ostree_boot_efi_path):
+    '''Recursively compare the files to upgrade'''
+    dcmp = filecmp.dircmp(esp_mount, ostree_boot_efi_path)
+    return dir_cmp(dcmp)
+
+
+def do_uefi_upgrade(esp_mount, ostree_boot_efi_path):
+    '''Actually start the UEFI upgrade copying the files'''
+    esp_mount_tmp = os.path.join(esp_mount, 'EFI.new')
+
+    print('[*] Updated:')
+    try:
+        file_list = distutils.dir_util.copy_tree(ostree_boot_efi_path,
+                                                 esp_mount_tmp)
+        os.sync()
+
+        for f_source in file_list:
+            f_rp = os.path.relpath(f_source, esp_mount_tmp)
+            f_dest = os.path.join(esp_mount, f_rp)
+
+            if not os.path.isfile(f_dest) or not filecmp.cmp(f_source, f_dest):
+                shutil.move(f_source, f_dest)
+                print(f_rp)
+    finally:
+        if os.path.exists(esp_mount_tmp):
+            distutils.dir_util.remove_tree(esp_mount_tmp)
+
+
+def start_uefi_upgrade():
+    '''Start the UEFI files upgrade'''
+    esp_mount = '/boot/efi'
+    ostree_boot_efi_path = os.path.join(cmdline['ostree'], 'boot/efi')
+
+    if not os.path.isdir(esp_mount):
+        raise BootUpgradeException('[!] ESP mountpoint',
+                                   esp_mount, 'does not exist')
+
+    disk = get_disk()
+    esp_partition = get_esp(disk)
+
+    with mount(esp_partition.device, esp_mount) as esp_mount:
+        if not check_for_upgrade(esp_mount, ostree_boot_efi_path):
+            print('[*] No UEFI files to upgrade')
+            return
+
+        if not check_esp_space(esp_mount, ostree_boot_efi_path):
+            raise BootUpgradeException('[!] Not enough space in the ESP')
+
+        do_uefi_upgrade(esp_mount, ostree_boot_efi_path)
+
+
+def start_conf_upgrade():
+    '''Start the GRUB configuration upgrade'''
+    conf_path = '/boot/grub/grub.cfg'
+    conf_path_tmp = conf_path + '.tmp'
+    conf_path_new = '/usr/lib/grub/conf/grub.cfg'
+
+    if filecmp.cmp(conf_path, conf_path_new):
+        print('[*] No configuration file to upgrade')
+        return
+
+    try:
+        shutil.copy2(conf_path_new, conf_path_tmp)
+        os.sync()
+
+        shutil.move(conf_path_tmp, conf_path)
+    finally:
+        if os.path.isfile(conf_path_tmp):
+            os.remove(conf_path_tmp)
+
+    print('[*] Updated:')
+    print(conf_path)
+
+
+if __name__ == '__main__':
+    cmdline = parse_cmdline()
+
+    if 'endless.image.device' in cmdline:
+        print('[!] Dual boot detected. Bailing out.')
+        exit(0)
+
+    if 'ostree' not in cmdline:
+        print('[!] Not an ostree system?')
+        exit(0)
+
+    if not os.path.exists('/sys/firmware/efi'):
+        print('[!] BIOS upgrade not yet implemented')
+        exit(0)
+
+    print('==> Starting UEFI/configuration upgrade...')
+    start_uefi_upgrade()
+    start_conf_upgrade()
```

Comment:
To narrow the window of time in which a power failure would leave the system unbootable, it might be better to interleave these two:

1. copy new config file to /boot/grub/grub.cfg.new
2. copy new ESP files to $ESP/EFI.new
3. `os.sync()`
4. rename everything into place, if changed
5. `os.sync()`
6. delete any left-over files

(There will be a lot of disk IO happening because of other post-update jobs so we'd expect `sync()` to be slow.)